File: video3-open-close/java/NotificationService2.java
public class NotificationService2 {

    public static void main(String[] args) {
        NotificationService2 ns = new NotificationService2();
        ns.sendNotification(new EmailService(), "helo");
        ns.sendNotification(new MessageService(), "hello");
        ns.sendNotification(new WechatService(), "hello");
    }

    public void sendNotification(NotificationService ns, String info) {
        ns.sendNotification(info);
    }
}

interface NotificationService {
    public void sendNotification(String info);
}

class EmailService implements NotificationService {
    public void sendNotification(String info) {
        System.out.printf("Email: %s\n", info);
    }
}

class MessageService implements NotificationService {
    public void sendNotification(String info) {
        System.out.printf("Message: %s\n", info);
    }
}

class WechatService implements NotificationService {
    public void sendNotification(String info) {
        System.out.printf("Wechat: %s\n", info);
    }
}

File: v3_open_close/python/notification_service2.py
class NotificationService:

    def send_notification(self, info):
        pass

class EmailService(NotificationService):

    def send_notification(self, info):
        print(f'Email: {info}')

class MessageService(NotificationService):

    def send_notification(self, info):
        print(f'Message: {info}')
    
class WeChatService(NotificationService):

    def send_notification(self, info):
        print(f'WeChat: {info}')

class NotificationService2:

    def send_notification(self, notification_service, info):
        notification_service.send_notification(info)

ns = NotificationService2()
ns.send_notification(EmailService(), 'hello')
ns.send_notification(MessageService(), 'hello')
ns.send_notification(WeChatService(), 'hello')

